resource_monitors: Add cgroupsv2 support for CPU Resource Monitor

rajeshetty87 opened the pull request:
**Commit Message**: Add cgroupsv2 support for CPU Resource Monitor 

**Additional Description:**

Currently CPU resource monitor exclusively works for cgroupsv1 format. When the envoy runs on instances that support cgrougsv2 we see the below error 
```
2025-09-18T22:20:21.751941Z	error	envoy misc external/envoy/source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:67	Can't open linux cpu allocated file /sys/fs/cgroup/cpu/cpu.shares	thread=22
2025-09-18T22:20:21.751984Z	info	envoy main external/envoy/source/server/overload_manager_impl.cc:754	Failed to update resource envoy.resource_monitors.cpu_utilization: Can't open file to read CPU utilization	thread=22
```
This is due the difference in folder structure and values in cgroupv2 compared to cgroupsv1. More details can be found in issues [#40571](https://github.com/envoyproxy/envoy/issues/40571),  [#39978](https://github.com/envoyproxy/envoy/issues/39978)

cgroupsv1
```
$ ls /sys/fs/cgroup -1
blkio
cpu
cpu,cpuacct
cpuacct
cpuset
devices
freezer
hugetlb
memory
net_cls
net_cls,net_prio
net_prio
perf_event
pids
systemd
```
Cgroupsv2
```
$ ls /sys/fs/cgroup -1
cgroup.controllers
cgroup.events
cgroup.freeze
cgroup.kill
cgroup.max.depth
cgroup.max.descendants
cgroup.pressure
cgroup.procs
cgroup.stat
cgroup.subtree_control
cgroup.threads
cgroup.type
cpu.idle
cpu.max
cpu.max.burst
cpu.pressure
cpu.stat
cpu.weight
cpu.weight.nice
cpuset.cpus
cpuset.cpus.effective
cpuset.cpus.partition
cpuset.mems
cpuset.mems.effective
```

To calculate utilization we used the `usage_usec` in `cpu.stat` file across multiple intervals and normalized it to the groups CPU capacity. Unlike cgroupsv1 the usage values are in microseconds instead of milliseconds. The `cpu.max` values show `60000` that is ~600m which is current cpu limit assigned to the container. 

```
$ cd /sys/fs/cgroup
$ more cpuset.cpus.effective
0-15
$ more cpu.stat
usage_usec 1759327475
user_usec 1236034525
system_usec 523292950
core_sched.force_idle_usec 0
nr_periods 70619
nr_throttled 10445
throttled_usec 385185405
nr_bursts 0
burst_usec 0
$ more cpu.max
60000 100000
$

```
The goal of this change is to compute normalized CPU utilization — a value between 0.0 and 1.0, representing how much CPU capacity a cgroup is using relative to its assigned limits.

**Core Formulae**

1. _CPU Usage and Interval_
```
Δusage    = usage_usec(t2) − usage_usec(t1)    # microseconds
Δinterval = t2 − t1                            # seconds
```
Δusage: Change in total CPU time used between two timestamps
Δinterval: Real elapsed time between measurements

2. _Effective CPU Capacity_

Let:

N = number of CPUs available to the cgroup
(usually from cpuset.cpus.effective)
```
cpu.max = quota and period configuration, defined as:
cpu.max = <quota> <period>
```
If a quota applies:
```
Q = quota / period            # in CPU cores
C = min(N, Q)                 # effective capacity
```
Example:

cpu.max = 60000 / 1000000
Q = 60000 / 1000000 = 0.06
If no quota is set (i.e. cpu.max = max):

C = N

3. Normalized CPU Utilization
```
util_normalized = (Δusage / 1_000_000) / (Δinterval * C)
```
Where:
- Δusage / 1_000_000 converts microseconds to seconds
- The denominator (Δinterval * C) represents total available CPU time
- util_normalized is a fractional value (0.0 to 1.0)

Example 

Parameter | Description | Value
-- | -- | --
usage_usec(t1) | CPU time at start | 1,000,000 µs
usage_usec(t2) | CPU time at end | 2,200,000 µs
Δusage | 2,200,000 − 1,000,000 | 1,200,000 µs
Δinterval | Elapsed time | 2 s
C | Effective capacity | 2 CPUs

```
util_normalized = (1,200,000 / 1_000_000) / (2 * 2)
                 = 1.2 / 4
                 = 0.3
```
Normalized CPU Utilization = 0.3 (30%)


Risk Level: 
**Testing**: 
- Added comprehensive unit tests. 
 - We have an internal fork on which we updated the resource monitor with the same code tested it across below test cases. 
       - 1. Overload manager CPU Utilization Resource Monitor on AL2023 instances with CPU Limit
       - 2. Overload manager CPU Utilization Resource Monitor on AL2023 instances without CPU Limit
       - 3. Overload manager CPU Utilization Resource Monitor on Amazon Linux 2023 with different instances types 
       - 4. Overload manager CPU Utilization Resource Monitor on Amazon Linux 2 for cgroupsv2 backwards compatibility
       
Docs Changes: 
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Fixes [#40571](https://github.com/envoyproxy/envoy/issues/40571),  [#39978](https://github.com/envoyproxy/envoy/issues/39978)
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.h:
```diff
@@ -29,13 +37,19 @@ class LinuxContainerCpuStatsReader : public CpuStatsReader {
       TimeSource& time_source,
       const std::string& linux_cgroup_cpu_allocated_file = LINUX_CGROUP_CPU_ALLOCATED_FILE,
-      const std::string& linux_cgroup_cpu_times_file = LINUX_CGROUP_CPU_TIMES_FILE);
+      const std::string& linux_cgroup_cpu_times_file = LINUX_CGROUP_CPU_TIMES_FILE,
```
I don't think there's a scenario where we'd use both cgroupv1 and cgroupv2 at the same time -- I'm not sure if it's even possible.

Can we do something similar to https://github.com/envoyproxy/envoy/blob/4c5d150710ba38add0acfcc2dff6a1edb2f363f1/source/extensions/resource_monitors/cgroup_memory/cgroup_memory_stats_reader.h#L79-L98 and have a seperate reader for the different implementations?

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -32,68 +36,219 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+  std::ifstream v2_stat_file(cgroupv2_cpu_stat_file_);
+  std::ifstream v2_max_file(cgroupv2_cpu_max_file_);
+  std::ifstream v2_effective_file(cgroupv2_cpu_effective_file_);
+
```
We could do all of this detection in the factory that decides which of the implementations to choose vs always trying cgroupv2 files on a system using cgroupv1 inefficiently

This would reduce the size of this function significantly.

rajeshetty87 commented:
@KBaichoo updated 
* refactored the code to use a factory pattern 
* moved utilization calculations to CpuTime object
* updated change log

Please take a look when you get a chance.

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -32,68 +38,241 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+
+  TRY_ASSERT_MAIN_THREAD {
+    const double cpu_allocated_value = std::stod(shares_result.value());
+    const double cpu_times_value = std::stod(usage_result.value());
```
can you use the absl numeric libs here: https://github.com/abseil/abseil-cpp/blob/5b1e199db4bc8f83f3d2dc1ced3dee7576ceedc2/absl/strings/numbers.h#L65-L88

this way we can avoid needing a try-catch and here as it was prior for cgroupv1

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -32,68 +38,241 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
-                   linux_cgroup_cpu_allocated_file_);
-    return {false, 0, 0};
+  const size_t dash_pos = range_token.find('-');
+  TRY_ASSERT_MAIN_THREAD {
```
likewise can we avoid try/catch by using the absl lib

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:
```diff
@@ -40,20 +44,24 @@ void CpuUtilizationMonitor::updateResourceUsage(Server::ResourceUpdateCallbacks&
+
+  // Calculate utilization (encapsulates cgroup v1 vs v2 differences and validation)
+  double current_utilization;
+  TRY_ASSERT_MAIN_THREAD {
```
I think we can get rid of the exception here if the `calculateUtilization` returned an `absl::Status` as well

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:
```diff
@@ -40,20 +44,24 @@ void CpuUtilizationMonitor::updateResourceUsage(Server::ResourceUpdateCallbacks&
+  ENVOY_LOG_MISC(
+      trace, "CPU utilization: {}, work_time={}, total_time={}, effective_cores={} (cgroup v{})",
+      current_utilization, cpu_times.work_time, cpu_times.total_time, cpu_times.effective_cores,
+      cpu_times.is_cgroup_v2 ? 2 : 1);
```
Do you think we'd need cgroup v1 vs cgroupv2 every log? Or would this info mostly be helpful when we first create the monitor (given that it wouldn't change)?

rajeshetty87 replied:
Simplified the trace log.

KBaichoo reviewed (commented):
thank you for working on this; this is looking better!

/wait

cancecen commented:
@rajeshetty87 As I understand, you are using the container CPU resource monitor to measure the utilization of the Envoy sidecar itself, is that correct? In my company we are using the "Linux" CPU resource monitor to measure the utilization of the whole VM, including the workload sits next to the Envoy process, which helps because we also want to take into the workload utilization itself. In containerized world, this doesn't work - so I am wondering what your use-case is and how your setup is like.

rajeshetty87 commented on source/extensions/resource_monitors/cpu_utilization/cpu_stats_reader.h:
```diff
@@ -18,9 +21,39 @@ namespace CpuUtilizationMonitor {
+   * @return The calculated CPU utilization as a fraction (0.0 to 1.0).
+   * @throw EnvoyException if the delta values are invalid (negative work or non-positive total).
+   */
+  double calculateUtilization(const CpuTimes& previous_cpu_times) const {
```
Understood, created specific implementations for Linux. CgroupsV1 and Cgroupsv2

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/cpu_stats_reader.h:
```diff
@@ -18,16 +23,25 @@ namespace CpuUtilizationMonitor {
+   * @return StatusOr containing utilization value (0.0 to 1.0) on success,
+   *         or InvalidArgumentError if calculation fails.
+   */
+  virtual absl::StatusOr<double> getUtilization() PURE;
```
I think now that we have this exposed, we no longer need `getCpuTimes` in public interface e.g. it can be private or removed and kept as an implementation detail for the classes since our user of this class just calls `getUtilization` directly.

rajeshetty87 replied:
Done.

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/cpu_utilization_monitor.cc:
```diff
@@ -30,39 +34,31 @@ CpuUtilizationMonitor::CpuUtilizationMonitor(
-                 previous_cpu_times_.work_time, cpu_times.work_time, previous_cpu_times_.total_time,
-                 cpu_times.total_time);
+
+  double current_utilization = utilization_result.value();
```
const

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -32,68 +40,336 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+      work_over_period_seconds / (total_over_period_seconds * current_cpu_times.effective_cores);
+
+  // Clamp to [0.0, 1.0]
+  const double clamped_utilization = std::clamp(utilization, 0.0, 1.0);
```
Can you help me understand what makes this utilization need clamping vs the other ones?

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -32,68 +40,336 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+  if (N <= 0) {
+    ENVOY_LOG(error, "No CPUs found in {}", effective_path_);
+    return {false, true, 0, 0, 0};
+  }
```
I _think_ this needs to handle ranges such as:
```
0-4,6,8-10
```

from 
https://docs.kernel.org/admin-guide/cgroup-v2.html#:~:text=cat%20cpuset.cpus-,0%2D4%2C6%2C8%2D10,-An%20empty%20value

rajeshetty87 replied:
Thanks for checking this, I totally missed this one. All the our setup runs on AWS AMI's and there seems to be no "parent" enforcement so the cpuset.cpus.effective always showed the values in the range 0-3, 0-7,0-15,0-31.  
Let me fix this code to handle all the robust cases.

rajeshetty87 replied:
The method is now updated to support mixed ranges

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -32,68 +40,336 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+  int N = 0;
+  std::string range_token = effective_result.value();
+  // Trim whitespace
+  range_token.erase(range_token.find_last_not_of(" \n\r\t") + 1);
```
prefer just using absl::StripTrailingAsciiWhitespace https://github.com/abseil/abseil-cpp/blob/6e0a3164d29a00f35263d29c03ef44cd2d95309f/absl/strings/ascii.h#L251

rajeshetty87 replied:
Done

KBaichoo commented on test/extensions/resource_monitors/cpu_utilization/cpu_paths_test.cc:
```diff
@@ -0,0 +1,571 @@
+
+// =============================================================================
+// Idempotency Tests - Verify methods return consistent values
+// =============================================================================
```
I noticed that some of the new test cases seem to be auto-generated and might not be testing real-world scenarios or critical paths.

To keep our test suite maintainable and fast, we try to focus on tests that cover:
- Core logic
- Known bug reproduction.
- Likely edge cases.

Could you please review the suite and trim it down to only the essential tests?

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -51,49 +61,320 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+
+  ENVOY_LOG(trace, "cgroupv2 usage_usec: {}", usage_usec);
+  ENVOY_LOG(trace, "cgroupv2 effective_cores: {}", effective_cores);
+  ENVOY_LOG(trace, "cgroupv2 current_time: {}", current_time);
```
likewise

KBaichoo commented on source/extensions/resource_monitors/cpu_utilization/linux_cpu_stats_reader.cc:
```diff
@@ -51,49 +61,320 @@ CpuTimes LinuxCpuStatsReader::getCpuTimes() {
+    }
+    const double q_cores = static_cast<double>(quota) / static_cast<double>(period);
+    effective_cores = std::min(static_cast<double>(N), q_cores);
+  }
```
likewise for effective cores

KBaichoo commented on test/extensions/resource_monitors/cpu_utilization/config_test.cc:
```diff
@@ -48,10 +72,85 @@ TEST(CpuUtilizationMonitorFactoryTest, CreateContainerCPUMonitor) {
+    // If no cgroup files exist (macOS, non-containerized), exception is expected
+    // This is acceptable
+  }
+}
```
can you use `#if defined(__linux__) `

for the parts that are linux checking its `EXPECT_NE `

and for other OSes that it throws have it else into a 

`EXPECT_THROW(...)`

KBaichoo commented on test/extensions/resource_monitors/cpu_utilization/config_test.cc:
```diff
@@ -48,10 +72,85 @@ TEST(CpuUtilizationMonitorFactoryTest, CreateContainerCPUMonitor) {
+
+  // Factory will either create a monitor (if cgroup files exist) or throw (if they don't)
+  // Test both behaviors
+  try {
```
For tests like this that are only applicable on linux, we can just run them on linux using the above technique.

KBaichoo commented on test/extensions/resource_monitors/cpu_utilization/cpu_paths_test.cc:
```diff
@@ -0,0 +1,571 @@
+  const auto usage_path = CpuPaths::V1::getUsagePath();
+  EXPECT_TRUE(usage_path.find("/sys/fs/cgroup/cpuacct") == 0);
+  EXPECT_TRUE(usage_path.find("cpuacct.usage") != std::string::npos);
+}
```
it seems like this would be covered by `GetSharesPathReturnsCorrectPath` and the other tests?

If so can we remove this

KBaichoo commented on test/extensions/resource_monitors/cpu_utilization/cpu_paths_test.cc:
```diff
@@ -0,0 +1,571 @@
+}
+
+// Test that paths are properly constructed (not empty or malformed)
+TEST(CpuPathsV2Test, PathsAreWellFormed) {
```
likewise to above

KBaichoo commented on test/extensions/resource_monitors/cpu_utilization/cpu_paths_test.cc:
```diff
@@ -0,0 +1,571 @@
+  EXPECT_FALSE(CpuPaths::isV2(mock_fs));
+}
+
+TEST(CpuPathsDetectionTest, NeitherV1NorV2WhenNoFilesExist) {
```
Can you help me understand what this adds over the independent tests that show these are set to false independently?

Otherwise can we remove

KBaichoo commented on test/extensions/resource_monitors/cpu_utilization/cpu_paths_test.cc:
```diff
@@ -0,0 +1,571 @@
+  const auto effective3 = CpuPaths::V2::getEffectiveCpusPath();
+  EXPECT_EQ(effective1, effective2);
+  EXPECT_EQ(effective2, effective3);
+}
```
These seems not helpful given the implementation -- are we expecting them to somehow be non-idempotent?

KBaichoo reviewed (commented):
Thanks again for working on this.

/wait
